Modules/_decimal/_decimal.c

```diff
@@ -3672,34 +3678,46 @@ dec_as_long(PyObject *dec, PyObject *context, int round)
     }
 
     status = 0;
-    ob_digit = NULL;
-#if PYLONG_BITS_IN_DIGIT == 30
-    n = mpd_qexport_u32(&ob_digit, 0, PyLong_BASE, x, &status);
-#elif PYLONG_BITS_IN_DIGIT == 15
-    n = mpd_qexport_u16(&ob_digit, 0, PyLong_BASE, x, &status);
-#else
-    #error "PYLONG_BITS_IN_DIGIT should be 15 or 30"
-#endif
 
-    if (n == SIZE_MAX) {
-        PyErr_NoMemory();
+    int64_t val = mpd_qget_i64(x, &status);
+    if (!status) {
+        mpd_del(x);
+        return PyLong_FromInt64(val);
+    }
+
+    const PyLongLayout *layout = PyLong_GetNativeLayout();
+    const uint32_t base = (uint32_t)1 << layout->bits_per_digit;
+    size_t len = mpd_sizeinbase(x, base);
+    PyLongWriter *writer = PyLongWriter_Create(mpd_isnegative(x), len, &digits);
+    if (writer == NULL) {
         mpd_del(x);
         return NULL;
     }
 
-    if (n == 1) {
-        sdigit val = mpd_arith_sign(x) * ob_digit[0];
-        mpd_free(ob_digit);
+    status = 0;
+    /* mpd_sizeinbase can overestimate size by 1 digit, set it first to zero. */
+    if (base > UINT16_MAX) {
+        ((uint32_t *)digits)[len - 1] = 0;
+        n = mpd_qexport_u32((uint32_t **)&digits, len, base, x, &status);
+    }
+    else {
+        ((uint16_t *)digits)[len - 1] = 0;
+        n = mpd_qexport_u16((uint16_t **)&digits, len, base, x, &status);
+    }
+    /* mpd_qexport_*() functions above used with assumption, that no
+       resizing occur, i.e. len was obtained by a call to mpd_sizeinbase. */
+    assert(n == len || n == len - 1);
+
+    if (n == SIZE_MAX) {
+        PyErr_NoMemory();
+        PyLongWriter_Discard(writer);
         mpd_del(x);
-        return PyLong_FromLong(val);
+        return NULL;
     }
 
     assert(n > 0);
```

Comment:
This assertion looks redundant with `assert(n == len || n == len - 1);`.

Reply:
Yeah, I've moved above assert here and restored nonzero assert for x.